File: server/src/main/kotlin/dev/chargedbyte/projectbirdnest/model/Drone.kt
package dev.chargedbyte.projectbirdnest.model

data class Drone(
    val serialNumber: String,
    val model: String,
    val manufacturer: String,
    val mac: String,
    val ipv4: String,
    val ipv6: String,
    val firmware: String,
    val positionY: Double,
    val positionX: Double,
    val altitude: Double
)

File: client/src/model/Drone.ts
import * as t from "io-ts"

import { isLeft } from "fp-ts/Either"
import { PathReporter } from "io-ts/PathReporter"

export const DroneC = t.type({
  serialNumber: t.string,
  model: t.string,
  manufacturer: t.string,
  mac: t.string,
  ipv4: t.string,
  ipv6: t.string,
  firmware: t.string,
  positionY: t.number,
  positionX: t.number,
  altitude: t.number,
})

export type Drone = t.TypeOf<typeof DroneC>

export const parseDrone = (input: string): Drone | null => {
  const result = DroneC.decode(JSON.parse(input))
  if (isLeft(result)) {
    console.error(PathReporter.report(result))
    return null
  }
  return result.right
}
